components/mediators/cache/org.wso2.carbon.mediator.cache/src/main/java/org/wso2/carbon/mediator/cache/CacheMediator.java

```diff
@@ -0,0 +1,694 @@
+					response.setSOAP11(synCtx.isSOAP11());
+					Map<String, String> headers =
+							(Map) msgCtx.getProperty(org.apache.axis2.context.MessageContext.TRANSPORT_HEADERS);
+					String messageType = (String) msgCtx.getProperty(Constants.Configuration.MESSAGE_TYPE);
+					Map<String, Object> headerProperties = new HashMap<String, Object>();
+					headerProperties.put(org.apache.axis2.context.MessageContext.TRANSPORT_HEADERS, headers);
+					headerProperties.put(Constants.Configuration.MESSAGE_TYPE, messageType);
+					response.setHeaderProperties(headerProperties);
+				}
+
+			} catch (XMLStreamException e) {
+				handleException("Unable to set the response to the Cache", e, synCtx);
+			}
+
+			if (response.getTimeout() > 0) {
+				response.setExpireTimeMillis(System.currentTimeMillis() + response.getTimeout());
+			}
+
+			getMediatorCache().put(response.getRequestHash(), response);
+			// Finally, we may need to replicate the changes in the cache
+			Replicator.replicate(cfgCtx);
+		} else {
+			synLog.auditWarn("A response message without a valid mapping to the " +
+			                 "request hash found. Unable to store the response in cache");
+		}
+
+	}
+
+	/**
+	 * Processes a request message through the cache mediator. Generates the request hash and looks
+	 * up for a hit, if found; then the specified named or anonymous sequence is executed or marks
+	 * this message as a response and sends back directly to client.
+	 *
+	 * @param synCtx incoming request message
+	 * @param synLog the Synapse log to use
+	 * @return should this mediator terminate further processing?
+	 * @throws ClusteringFault if there is an error in replicating the cfgCtx
+	 */
+	private boolean processRequestMessage(MessageContext synCtx,
+	                                      SynapseLog synLog) throws ClusteringFault {
+
+		if (collector) {
+			handleException("Request messages cannot be handled in a collector cache", synCtx);
+		}
+
+		OperationContext opCtx = ((Axis2MessageContext) synCtx).getAxis2MessageContext().getOperationContext();
+		String requestHash = null;
+
+		try {
+			requestHash = digestGenerator.getDigest(((Axis2MessageContext) synCtx).getAxis2MessageContext());
+			synCtx.setProperty(CachingConstants.REQUEST_HASH, requestHash);
+		} catch (CachingException e) {
+			handleException("Error in calculating the hash value of the request", e, synCtx);
+		}
+
+		if (synLog.isTraceOrDebugEnabled()) {
+			synLog.traceOrDebug("Generated request hash : " + requestHash);
+		}
+
+		RequestHash hash = new RequestHash(requestHash);
+		CachableResponse cachedResponse = getMediatorCache().get(requestHash);
+		org.apache.axis2.context.MessageContext msgCtx = ((Axis2MessageContext) synCtx).getAxis2MessageContext();
+		opCtx.setProperty(CachingConstants.REQUEST_HASH, requestHash);
+
+		byte[] responseEnvelop;
+		Map<String, Object> headerProperties;
+		if (cachedResponse != null && (responseEnvelop = cachedResponse.getResponseEnvelope()) != null) {
```

Comment:
We should be using java.utils.Arrays.equals() in place of using == when comparing the two byte arrays.